In [7]:
import pandas as pd

In [8]:
df=pd.read_csv("/home/user/Downloads/fertilizer_recommendation_cleaned.csv")

CLASSIFICATION

`Logistic Regression Algorithm`

Logistic Regression calculates a weighted sum of input features, applies a sigmoid function to convert it into a probability, and then classifies the output based on a threshold (usually 0.5).

In [9]:

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
#choosing only suitable  column values
df = df[(df['Soil_pH'] >= 0) & (df['Soil_pH'] <= 14)]
#convert the text columns into numbers 
# because Machine Learning models cannot understand text (categorical data) — they only understand numbers.
le = LabelEncoder()
categorical_cols = df.select_dtypes(include=['object']).columns
for col in categorical_cols:
    df[col] = le.fit_transform(df[col])
#storing the input columns
X = df.drop('Recommended_Fertilizer', axis=1)
#storing the output column
y = df['Recommended_Fertilizer']
# train-test split
#here we are choose 80% data should be trained remaining will be used for testing
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
#here we are scaling because the values of the all features(input columns) are in same range then no one columns dominate the other 
#column value
scaler = StandardScaler()
#fit-learns 
#transform-apply on the data
X_train = scaler.fit_transform(X_train)
#Uses same mean & std from training data
X_test = scaler.transform(X_test)
#Train Logistic Regression Model
#max_iter=It means maximum number of iterations (steps) the Logistic Regression algorithm will take to learn.
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)
#predict
y_pred = model.predict(X_test)
#Evaluation
print("Accuracy:", accuracy_score(y_test, y_pred))

print("\nClassification Report:\n")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.533256880733945

Classification Report:

              precision    recall  f1-score   support

           0       0.48      0.18      0.26       179
           1       0.46      0.61      0.52       521
           2       0.61      0.44      0.51       245
           3       0.00      0.00      0.00       106
           4       0.00      0.00      0.00        30
           5       0.60      0.87      0.71       528
           6       0.50      0.10      0.17       135

    accuracy                           0.53      1744
   macro avg       0.38      0.31      0.31      1744
weighted avg       0.49      0.53      0.48      1744


Confusion Matrix:

[[ 32  97   2   1   0  46   1]
 [ 20 318  56   1   0 122   4]
 [  4  73 109   4   0  50   5]
 [  2  63   1   0   0  40   0]
 [  2  19   0   0   0   7   2]
 [  6  52  10   1   0 457   2]
 [  1  76   2   3   0  39  14]]


/home/user/Dsp_lab/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/home/user/Dsp_lab/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/home/user/Dsp_lab/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


`classfication_report`

precision:Out of predicted values, how many are correct?
Precision = TP / (TP + FP)

recall:Out of actual values, how many did model find? recall=TP/(TP/NP)

F1-SCORE:Balance between precision and recall


F1 = 2 * (Precision * Recall) / (Precision + Recall)

support:Number of actual samples in each class

We need a single summary value
That’s where macro avg and weighted avg come in

Macro Avg = (Score₁ + Score₂ + ... + Scoreₙ) / n

`CONFUSION MATRIX`

A confusion matrix is a table that compares actual and predicted values to evaluate the performance of a classification model and identify correct and incorrect predictions.

Rows = Actual values
Columns = Predicted values

`K-NEAREST NEIGHBOUR ALGORITHM`

K-Nearest Neighbors (KNN) works by finding the k closest data points to a new input based on distance and assigning the class that is most common among those neighbors.

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix
df=df[(df['Soil_pH'])>=0 & (df['Soil_pH']<=14)]
y = df['Recommended_Fertilizer']
X = df.drop('Recommended_Fertilizer', axis=1)
#Finds all categorical columns
#Converts them into multiple columns
X = pd.get_dummies(X)
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)
scaler=StandardScaler()
X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)
#selecting no of neighbors 9 for accuracy 
model=KNeighborsClassifier(n_neighbors=9)
model.fit(X_train,y_train)
y_pred=model.predict(X_test)
print("Acuuracy:",accuracy_score(y_test,y_pred))
print("\nClassification Report:\n")
print(classification_report(y_test,y_pred))
print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test,y_pred))



Acuuracy: 0.36811926605504586

Classification Report:

              precision    recall  f1-score   support

           0       0.28      0.23      0.25       179
           1       0.35      0.51      0.42       521
           2       0.34      0.22      0.26       245
           3       0.30      0.10      0.15       106
           4       0.00      0.00      0.00        30
           5       0.44      0.50      0.47       528
           6       0.18      0.04      0.07       135

    accuracy                           0.37      1744
   macro avg       0.27      0.23      0.23      1744
weighted avg       0.34      0.37      0.34      1744


Confusion Matrix:

[[ 41  81  11   0   1  41   4]
 [ 34 266  42  10   0 163   6]
 [ 19  95  53   5   0  68   5]
 [ 13  44   7  11   0  29   2]
 [  6  15   3   0   0   5   1]
 [ 32 177  35   8   2 265   9]
 [  4  77   7   3   1  37   6]]


`Support Vector Machines(SVM) Algorithm`

Support Vector Machine (SVM) works by finding the optimal boundary (hyperplane) that best separates different classes by maximizing the margin between them, and it can handle non-linear data using kernel functions.

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
df = df[(df['Soil_pH'] >= 0) & (df['Soil_pH'] <= 14)]
y = df['Recommended_Fertilizer']
X = df.drop('Recommended_Fertilizer', axis=1)
X = pd.get_dummies(X)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)
# kernel helps svm handle non-linear data by tranforming it into a higher dimension where it becomes separable
#rbf-radial basis function is the default kernel in svm that maps data into a higher-dimensional space to handle
#non-linear classification problems
# rbf = best default
model = SVC(kernel='rbf')   
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n")
print(classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.5951834862385321

Classification Report:

              precision    recall  f1-score   support

           0       0.52      0.31      0.39       179
           1       0.56      0.76      0.64       521
           2       0.64      0.46      0.53       245
           3       0.48      0.14      0.22       106
           4       0.00      0.00      0.00        30
           5       0.64      0.82      0.72       528
           6       0.56      0.21      0.30       135

    accuracy                           0.60      1744
   macro avg       0.49      0.39      0.40      1744
weighted avg       0.58      0.60      0.56      1744


Confusion Matrix:

[[ 56  67  10   4   0  41   1]
 [ 11 394  25   2   0  86   3]
 [ 10  60 112   2   0  54   7]
 [  9  43   8  15   0  25   6]
 [  7  15   0   1   0   6   1]
 [  7  75   9   0   0 433   4]
 [  7  54  11   7   0  28  28]]


/home/user/Dsp_lab/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/home/user/Dsp_lab/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/home/user/Dsp_lab/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


`Decision Tree Algorithm`

Decision Tree works by splitting the data into smaller subsets based on feature conditions, forming a tree-like structure where each path leads to a final class prediction.

In [12]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
df = df[(df['Soil_pH'] >= 0) & (df['Soil_pH'] <= 14)]
y = df['Recommended_Fertilizer']
X = df.drop('Recommended_Fertilizer', axis=1)
X = pd.get_dummies(X)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n")
print(classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.5963302752293578

Classification Report:

              precision    recall  f1-score   support

           0       0.37      0.36      0.36       179
           1       0.71      0.69      0.70       521
           2       0.58      0.58      0.58       245
           3       0.34      0.30      0.32       106
           4       0.10      0.13      0.12        30
           5       0.73      0.74      0.74       528
           6       0.32      0.34      0.33       135

    accuracy                           0.60      1744
   macro avg       0.45      0.45      0.45      1744
weighted avg       0.60      0.60      0.60      1744


Confusion Matrix:

[[ 64  31  22  14   8  25  15]
 [ 29 360  31  10   4  64  23]
 [ 18  31 141   8   5  24  18]
 [ 15  13   7  32   6  14  19]
 [  7   6   3   1   4   4   5]
 [ 27  51  20  13   5 393  19]
 [ 12  18  21  17   7  14  46]]


`Random Forest Algorithm`

Random Forest works by building multiple decision trees on different subsets of the data and combining their predictions (majority voting) to improve accuracy and reduce overfitting.

In [13]:

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
df = df[(df['Soil_pH'] >= 0) & (df['Soil_pH'] <= 14)]
y = df['Recommended_Fertilizer']
X = df.drop('Recommended_Fertilizer', axis=1)
X = pd.get_dummies(X)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
#n_estimators:Number of trees in the Random Forest
#max_depth:maximum number of levels each decision tree can grow
model = RandomForestClassifier(
    n_estimators=100,      
    max_depth=None,        
    random_state=42
)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n")
print(classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.6955275229357798

Classification Report:

              precision    recall  f1-score   support

           0       0.50      0.66      0.57       179
           1       0.75      0.82      0.78       521
           2       0.67      0.65      0.66       245
           3       0.52      0.36      0.42       106
           4       0.00      0.00      0.00        30
           5       0.87      0.78      0.82       528
           6       0.39      0.45      0.42       135

    accuracy                           0.70      1744
   macro avg       0.53      0.53      0.52      1744
weighted avg       0.69      0.70      0.69      1744


Confusion Matrix:

[[118  25  15   5   0   6  10]
 [ 33 425  28   2   0  21  12]
 [ 22  21 159   7   0  12  24]
 [ 16  16   4  38   0  10  22]
 [ 10   5   4   0   0   4   7]
 [ 20  51  18   5   0 412  22]
 [ 15  22  10  16   0  11  61]]


/home/user/Dsp_lab/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/home/user/Dsp_lab/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/home/user/Dsp_lab/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


`XGBOOST Algorithm`

XGBoost:XGBoost is a powerful ensemble learning algorithm based on gradient boosting.
It builds a series of decision trees, where each tree tries to correct the errors of the previous ones.
Known for its high accuracy, speed, and ability to handle large datasets.

In [ ]:

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from xgboost import XGBClassifier
df = df[(df['Soil_pH'] >= 0) & (df['Soil_pH'] <= 14)]
y = df['Recommended_Fertilizer']
X = df.drop('Recommended_Fertilizer', axis=1)
X = pd.get_dummies(X)
#LabelEncoder is a tool from scikit-learn (sklearn.preprocessing) used to convert categorical labels into numerical values.
le = LabelEncoder()
y = le.fit_transform(y)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)
#n_estimators=300 => The number of trees in the model.
#learning_rate=0.1 => Controls how much each tree contributes to the final prediction.
#max_depth=6 => Maximum depth of each tree
#subsample=0.8 => Fraction of rows randomly selected for each tree.
#colsample_bytree=0.8 => Fraction of features (columns) used for each tree.
#n_jobs: Tells XGBoost to use all available CPU cores for faster training.
#n_jobs=1 => single core (slow).
#n_jobs=-1 => maximum parallelism.
model = XGBClassifier(
    n_estimators=300,       
    learning_rate=0.1,      
    max_depth=6,           
    subsample=0.8,          
    colsample_bytree=0.8,   
    random_state=42,
    n_jobs=-1
)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.676605504587156

Classification Report:
               precision    recall  f1-score   support

           0       0.47      0.52      0.50       174
           1       0.72      0.81      0.76       509
           2       0.67      0.62      0.64       246
           3       0.50      0.41      0.45       110
           4       0.33      0.03      0.06        32
           5       0.81      0.80      0.80       543
           6       0.37      0.36      0.36       130

    accuracy                           0.68      1744
   macro avg       0.55      0.51      0.51      1744
weighted avg       0.67      0.68      0.67      1744


Confusion Matrix:
 [[ 91  32  12  11   0  16  12]
 [ 15 412  24  10   0  33  15]
 [ 17  30 152   4   0  22  21]
 [ 12  19   7  45   0  14  13]
 [ 15   4   1   0   1   3   8]
 [ 16  55  21   7   0 432  12]
 [ 26  18   9  13   2  15  47]]


`ANALYSIS`

1.Logistic Regression (0.53):
Performs poorly because our data is non-linear and multi-class.
Linear boundaries cannot capture complex relationships.

2.KNN (0.36):
Not Better than Logistic Regression.
Uses distance to classify, so it captures non-linear patterns.
Sensitive to feature scaling.

3.SVM (0.59):
Handles non-linear data with the RBF kernel.
More accurate than KNN because it finds optimal hyperplanes.

4.Decision Tree (0.59)
Captures non-linear relationships without scaling.
May overfit if depth is unlimited, but still better than simpler models.

5.Random Forest (0.69)
Best model overall.
Combines multiple decision trees to reduce overfitting.
Handles class imbalance and complex patterns well.

6.XGBoost (0.67):
Captures complex non-linear relationships and interactions. Handles class imbalance well, uses boosting to reduce errors from weak learners, and is robust against overfitting with proper regularization.

`CONCLUSION`


`Random Forest` is the most suitable model for this dataset.
Accuracy increases progressively with more sophisticated models: Logistic Regression -> KNN -> SVM -> Decision Tree -> Random Forest.
This shows that our data is multi-class, non-linear, and has complex feature interactions, which simpler linear models cannot capture.